c/test/cpp/test_sbp_stdio.cc

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (C) 2020 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <gtest/gtest.h>
+
+#include <libsbp/cpp/sbp_stdio.h>
+#include <libsbp/cpp/message_handler.h>
+
+namespace {
+
+struct SbpHeaderParams {
+  uint16_t sender_id;
+  uint16_t msg_type;
+  uint8_t payload_len;
+  msg_obs_t payload;
+};
+
+class MsgObsHandler : private sbp::MessageHandler<msg_obs_t> {
+public:
+  MsgObsHandler(sbp::State *state) : sbp::MessageHandler<msg_obs_t>(state), state_(state) {}
+
+  void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const msg_obs_t &msg) {
+    header_params_.sender_id = sender_id;
+    header_params_.msg_type = SBP_MSG_OBS;
+    header_params_.payload_len = message_length;
+    header_params_.payload = msg;
+  }
+
+  void write_message() const {
+    state_->send_message(header_params_.msg_type, header_params_.sender_id,
+                         header_params_.payload_len,
+                         reinterpret_cast<const uint8_t *>(&header_params_.payload));
+  }
+
+  SbpHeaderParams get_header_params() {
+    return header_params_;
+  }
+
+private:
+  SbpHeaderParams header_params_;
+  sbp::State *state_;
+};
+
+
+class SbpStdioTest : public ::testing::Test {
+protected:
+  int num_entries_in_file(const std::string &input_file) {
+    sbp::SbpFileReader reader = sbp::SbpFileReader(input_file);
+    sbp::State state;
+    state.set_reader(&reader);
+    MsgObsHandler handler(&state);
+
+    int count = 0;
+    while (true) {
+      s8 status = state.process();
+      if (status < SBP_OK) {
+        break;
+      }
+      if (status == SBP_OK_CALLBACK_EXECUTED) {
+        count++;
+      }
+    }
+    return count;
+  }
+
+  void write_to_file(const std::string &input_file, const std::string &output_file) {
+    sbp::SbpFileReader reader = sbp::SbpFileReader(input_file);
+    sbp::SbpFileWriter writer = sbp::SbpFileWriter(output_file);
+    sbp::State state;
+    state.set_reader(&reader);
+    state.set_writer(&writer);
+    MsgObsHandler handler(&state);
+
+    while (true) {
+      s8 status = state.process();
+      if (status < SBP_OK) {
+        break;
+      }
+      if (status == SBP_OK_CALLBACK_EXECUTED) {
+        handler.write_message();
+        handler.write_message();
+        handler.write_message();
+      }
+    }
+  }
+};
+
+
+TEST_F(SbpStdioTest, ReadsSbpFiles) {
+  EXPECT_EQ(num_entries_in_file("gnss_data.sbp"), 3);
```

Comment:
It looks like there's already a `test_data` directory already in the root of the repo. Could we reuse one of the files there to test this?

Reply:
So I looked into reusing one of the files there. Unfortunately, both the `short.sbp` and `long.sbp` files gave me an error when I read them in. This looks like a data error, not a problem with my code, as `sbp2json` also gave me this problem:
```
(py3) laykuanloh@Lays-MBP ~/Desktop$ cat long.sbp | sbp2json
Traceback (most recent call last):
  File "/Users/laykuanloh/py3/bin/sbp2json", line 6, in <module>
    module_main()
  File "/Users/laykuanloh/py3/lib/python3.8/site-packages/sbp/sbp2json.py", line 235, in module_main
    sbp_main(args)
  File "/Users/laykuanloh/py3/lib/python3.8/site-packages/sbp/sbp2json.py", line 211, in sbp_main
    m = msg_nojit.SBP.unpack(b)
  File "/Users/laykuanloh/py3/lib/python3.8/site-packages/sbp/msg.py", line 233, in unpack
    assert p.preamble == SBP_PREAMBLE, "Invalid preamble 0x%x." % p.preamble
AssertionError: Invalid preamble 0x0.
```

Reply:
This is bizarre, the data files there work fine on my machine (and on the CI for sbp2json).

Reply:
Talked, there's a bug in Python 3.8 working with sbp2json. Filed as https://swift-nav.atlassian.net/browse/INFRA-114.

Unfortunately the files in `test_data` still don't work - nothing gets read in at all. At this point I think its better to keep the files I added instead, which have the added benefit of allowing user to easily make changes to the test data if they desire.